File: spring-web-java/src/main/java/spring/web/java/infrastructure/logging/HttpLog.java
package spring.web.java.infrastructure.logging;

import java.nio.charset.StandardCharsets;
import java.util.Enumeration;

import org.springframework.http.HttpMethod;
import org.springframework.web.util.ContentCachingRequestWrapper;
import org.springframework.web.util.ContentCachingResponseWrapper;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;

import lombok.SneakyThrows;
import lombok.extern.slf4j.Slf4j;

@Slf4j
public record HttpLog(
	String transactionId,
	HttpMethod httpMethod,
	String requestUri,
	String remoteAddr,
	Double elapsedTime,
	String requestHeader,
	String requestParam,
	String requestBody,
	String responseBody
) {
	public HttpLog(
		String transactionId,
		ContentCachingRequestWrapper requestWrapper,
		ContentCachingResponseWrapper responseWrapper,
		Double elapsedTime
	) {
		this(
			transactionId,
			HttpMethod.valueOf(requestWrapper.getMethod()),
			requestWrapper.getRequestURI(),
			requestWrapper.getRemoteAddr(),
			elapsedTime,
			getHeader(requestWrapper),
			getParameter(requestWrapper),
			convertByteArrayToString(requestWrapper.getContentAsByteArray()),
			convertByteArrayToString(responseWrapper.getContentAsByteArray())
		);
	}

	private static String getHeader(ContentCachingRequestWrapper requestWrapper) {
		StringBuilder builder = new StringBuilder();
		Enumeration<String> headerNames = requestWrapper.getHeaderNames();

		while (headerNames.hasMoreElements()) {
			String name = headerNames.nextElement();
			String value = requestWrapper.getHeader(name);
			builder.append("\n").append("  ").append(name).append(": ").append(value);
		}

		return builder.toString();
	}

	private static String getParameter(ContentCachingRequestWrapper requestWrapper) {
		StringBuilder builder = new StringBuilder();
		Enumeration<String> parameterNames = requestWrapper.getParameterNames();

		while (parameterNames.hasMoreElements()) {
			String name = parameterNames.nextElement();
			String value = requestWrapper.getParameter(name);
			builder.append("\n").append("  ").append(name).append(": ").append(value);
		}

		return builder.toString();
	}

	@SneakyThrows
	private static String convertByteArrayToString(byte[] content) {
		String body = "";
		if (content.length > 0) {
			body = new String(content, StandardCharsets.UTF_8);
		}

		return setPretty(body);
	}

	@SneakyThrows
	private static String setPretty(String body) {
		try {
			ObjectMapper objectMapper = new ObjectMapper();
			ObjectWriter objectWriter = objectMapper.writerWithDefaultPrettyPrinter();
			Object json = objectMapper.readValue(body, Object.class);
			return objectWriter.writeValueAsString(json);
		} catch (Exception e) {
			return body;
		}
	}

	public void log() {
		String message = """
			\n|>> REQUEST: %s %s (%s)
			|>> TRANSACTION_ID: %s
			|>> CLIENT_IP: %s
			|>> REQUEST_HEADER: %s
			|>> REQUEST_PARAMETER: %s
			|>> REQUEST_BODY: %s
			|>> RESPONSE_BODY: %s
			"""
			.formatted(
				httpMethod,
				requestUri,
				elapsedTime,
				transactionId,
				remoteAddr,
				requestHeader,
				requestParam,
				requestBody,
				responseBody
			);

		log.info(message);
	}
}

File: spring-web-kotlin/src/main/kotlin/spring/web/kotlin/infrastructure/logging/HttpLog.kt
package spring.web.kotlin.infrastructure.logging

import com.fasterxml.jackson.databind.ObjectMapper
import org.slf4j.LoggerFactory
import org.springframework.http.HttpMethod
import org.springframework.web.util.ContentCachingRequestWrapper
import org.springframework.web.util.ContentCachingResponseWrapper
import java.nio.charset.StandardCharsets

data class HttpLog(
    val transactionId: String,
    val httpMethod: HttpMethod,
    val requestUri: String,
    val remoteAddr: String,
    val elapsedTime: Double,
    val requestHeader: String,
    val requestParam: String,
    val requestBody: String,
    val responseBody: String
) {
    private val log = LoggerFactory.getLogger(HttpLog::class.java)

    constructor(
        transactionId: String,
        requestWrapper: ContentCachingRequestWrapper,
        responseWrapper: ContentCachingResponseWrapper,
        elapsedTime: Double
    ): this(
        transactionId = transactionId,
        httpMethod = HttpMethod.valueOf(requestWrapper.method),
        requestUri = requestWrapper.requestURI,
        remoteAddr = requestWrapper.remoteAddr,
        elapsedTime = elapsedTime,
        requestHeader = getHeader(requestWrapper),
        requestParam = getParameter(requestWrapper),
        requestBody = convertByteArrayToString(requestWrapper.contentAsByteArray),
        responseBody = convertByteArrayToString(responseWrapper.contentAsByteArray)
    )

    companion object {
        private fun getHeader(requestWrapper: ContentCachingRequestWrapper) =
            requestWrapper.headerNames.asSequence()
                .joinToString { "\n  ${it}: ${requestWrapper.getHeader(it)}" }

        private fun getParameter(requestWrapper: ContentCachingRequestWrapper) =
            requestWrapper.parameterNames.asSequence()
                .joinToString { "\n  ${it}: ${requestWrapper.getParameter(it)}" }

        private fun convertByteArrayToString(content: ByteArray) =
            if (content.isNotEmpty()) setPretty(String(content, StandardCharsets.UTF_8)) else ""

        private fun setPretty(body: String) =
            runCatching {
                val objectMapper = ObjectMapper()
                val objectWriter = objectMapper.writerWithDefaultPrettyPrinter()
                val json = objectMapper.readValue(body, Any::class.java)
                objectWriter.writeValueAsString(json)
            }.getOrDefault(body)
    }

    fun log() {
        val message = """
            |>> REQUEST: $httpMethod $requestUri ($elapsedTime)
            |>> TRANSACTION_ID: $transactionId
            |>> CLIENT_IP: $remoteAddr
            |>> REQUEST_HEADER: $requestHeader
            |>> REQUEST_PARAMETER: $requestParam
            |>> REQUEST_BODY: $requestBody
            |>> RESPONSE_BODY: $responseBody
        """.trimMargin()
            .let { "\n$it" }

        log.info(message)
    }
}
